src/khepri_lock.erl

```diff
@@ -0,0 +1,654 @@
+    acquire(StoreId, LockId, Scope, Options).
+
+-spec acquire(StoreId, LockId, Scope, Options) -> Ret when
+      StoreId :: khepri:store_id(),
+      LockId :: khepri_lock:lock_id(),
+      Scope :: khepri_lock:scope(),
+      Options :: khepri_lock:options(),
+      Ret :: khepri:ok(Lock) | khepri:error(),
+      Lock :: khepri_lock:lock().
+%% @doc Attempts to acquire a lock.
+%%
+%% This function attempts to acquire a lock, returning `{ok, Lock}' if it is
+%% successful. If the lock can't be acquired and there are retries remaining,
+%% this function will sleep at random increasing intervals. The first sleep
+%% is at most 1/4 of a second. The maximum value doubles with each failed
+%% attempt to acquire the lock up to a maximum of 8 seconds. The actual time
+%% slept is randomized with {@link random:uniform/1}.
+%%
+%% If the number of retries is exhausted, this function returns `{error,
+%% unavailable}'. The number of retries can be configured with the `retries'
+%% option in {@link options()}. By default, this value is `infinity'
+%% representing unlimited retries.
+%%
+%% This function will throw `{bad_options, Options}' if an invalid combination
+%% of options is passed. For example, setting the `recursive' option to `false'
+%% while providing a `group' is invalid and will cause this function to throw.
+%%
+%% @param StoreId the name of the Khepri store.
+%% @param LockId a term representing the lock.
+%% @param Scope the scope to which the lock should apply.
+%% @param Options options for acquiring the lock.
+%%
+%% @returns `{ok, Lock}' if the lock is successfully acquired or `{error,
+%% unavailable}' if the lock is held and the number of retries are exhausted,
+%% and `{error, Reason}' with any other reason if the acquisition failed for
+%% expected reasons.
+
+acquire(StoreId, LockId, Scope, Options)
+  when ?IS_STORE_ID(StoreId) andalso
+       ?IS_SCOPE(Scope) andalso
+       is_map(Options) ->
+    GroupId = group_id_option(Options),
+    ReleaseOnExit = release_on_exit_option(Options, Scope),
+    Retries = maps:get(retries, Options, infinity),
+    Lock = #khepri_lock{store_id = StoreId,
+                        lock_id = LockId,
+                        group_id = GroupId,
+                        scope = Scope,
+                        release_on_exit = ReleaseOnExit},
+    acquire_with_backoff(Lock, Retries, 1).
+
+acquire_with_backoff(Lock, Retries, Attempts) ->
+    case try_acquire(Lock) of
+        ok ->
+            {ok, Lock};
+        {error, unavailable} when Retries > 0 ->
+            backoff_sleep(Attempts),
+            acquire_with_backoff(Lock, decrement(Retries), Attempts + 1);
+        {error, _} = Error ->
+            Error
+    end.
+
+try_acquire(#khepri_lock{scope = local,
+                         store_id = StoreId,
+                         lock_id = LockId,
+                         group_id = GroupId,
+                         release_on_exit = ReleaseOnExit}) ->
```

Comment:
In other locking libraries, I saw the name "attempt" for this that I think is also good. What do you think? "try_acquire" also seems good

Reply:
I'm used to see "try_lock", but something based on "attempt" is good too.

Let me write them to see what it would look like:
```erlang
khepri_lock:try_acquire(...),

khepri_lock:attempt(...),

khepri_lock:try(...),
```

In fact, I like the last two. A small preference for "try" as it is shorter, but not a strong feeling.

Reply:
Ah actually `try` is a problem because of the `try` keyword. We could quote it (`khepri_lock:'try'(...)`) but it's a little less friendly to work with